xtractmime/__init__.py

```diff
@@ -45,19 +49,95 @@ def _is_match_mime_pattern(
     return True
 
 
-def _find_unknown_mimetype(input_bytes: bytes, sniff_scriptable: bool):
-    # TODO
-    pass
+def _find_unknown_mimetype(input_bytes: bytes, sniff_scriptable: bool, extra_types: Optional[Tuple[Tuple[bytes, bytes, Set[bytes], str], ...]]) -> str:
+    if sniff_scriptable:
+        return get_text_mime(input_bytes)
+
+    matched_type = get_extra_mime(input_bytes, extra_types=extra_types)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_image_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_audio_video_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_archive_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    if not contains_binary(input_bytes):
+        return "text/plain"
+    
+    return "application/octet-stream"
+
+
+def _sniff_mislabled_binary(input_bytes: bytes) -> str:
+    input_size = len(input_bytes)
+    
+    if input_size >= 2 and input_bytes.startswith((b"\xfe\xff",b"\xff\xfe",b"\xef\xbb\xbf")):
+        return "text/plain"
+    
+    if not contains_binary(input_bytes):
+        return "text/plain"
+    
+    return "application/octet-stream"
+
+
+def _sniff_mislabled_feed(input_bytes: bytes, supplied_type: Optional[Tuple[bytes]]) -> str:
+    input_size = len(input_bytes)
+    index = 0
+
+    if input_bytes[:3] == b"\xef\xbb\xbf":
+        index += 3
+
+    while index < input_size:
+        while True:
+            if not input_bytes[index:index+1]:
+                return supplied_type
+
+            if input_bytes[index:index+1] == b"<":
+                index += 1
+                break
+
+            if input_bytes[index:index+1] not in WHITESPACE_BYTES:
+                return supplied_type
+
+            index += 1
+    
+        while True:
+            if not input_bytes[index:index+1]:
+                return supplied_type
+
+            if input_bytes[index:index+3] == b"!--":
+                index += 3
+                while True:
+                    if not input_bytes[index:index+1]:
+                        return supplied_type
 
+                    if input_bytes[index:index+3] == b"-->":
+                        index += 3
+                        return supplied_type
```

Comment:
When the standard says “exit loops M and L”, I believe they mean to exit the 2 `while True` loops, but _not_ return. I believe the goal is to continue to the next iteration of the root `while index < input_size:` loop (hence the change to `index`, which would be pointless if there is a return afterwards).

You can usually [use a boolean parameter to handle that logic](https://stackoverflow.com/a/3357305/939364).

The algorithm aims to:
- Find the first `<` that is not part of an HTML comment (`<!-- … -->`), a markup declaration (`<! … >`), or a document type definition (`<? … ?>`).
- If it’s part of `<rss`, it’s `application/rss+xml`.
- If it’s part of `<feed`, it’s `application/atom+xml`.
- (put the complex rdf:RDF logic here :slightly_smiling_face: )